### DISCUSSION TITLE
Is there a strict mode for MockProvider

### DISCUSSION THREAD

**Discussion by bolemeus** [help-me-mom/ng-mocks#9448] [Q&A] [answered]
Before using this library I wrote all the mocks myself. Something like...
```
export class ExampleServiceMock implements Pick<ExampleService, keyof ExampleService> {
    method1(_arg: string): Observable<Example> {
      throw new Error('ExampleServiceMock.method1 is not mocked');
    }

    method2(_arg: string): Observable<Example> {
      throw new Error('ExampleServiceMock.method2 is not mocked');
    }
}
```

Is it possible to get the mocks to throw a similar Error when the code is calling a method that is not mocked explicitly?
Doing it like this always resulted in a clear error message when forgetting something simple in a test case.
The errors I get now when I forget something is like...
```
Can't read <field> of undefined
```

I'm looing for an easy way to get back to this behaviour, without resorting to provide the implementations for all the methods I'm mocking again.

**Comment by satanTime** [comment, marked as answer]
There is no separate `strict` option on `MockProvider`, but [`ngMocks.autoSpy`](https://ng-mocks.sudo.eu/extra/auto-spy/) is the intended extension point and can provide this behavior globally.

For Jasmine, put this in your test setup file:

```ts
ngMocks.autoSpy(label =>
  jasmine.createSpy(label).and.callFake(() => {
    throw new Error(`${label} is not mocked`);
  }),
);
```

For Jest:

```ts
ngMocks.autoSpy(label =>
  jest
    .fn(() => {
      throw new Error(`${label} is not mocked`);
    })
    .mockName(label),
);
```

The label already contains the class and method names, for example `ExampleService.method1`. Therefore, calling an unstubbed generated method fails immediately with:

```text
ExampleService.method1 is not mocked
```

Explicit `MockProvider` overrides are applied after these default methods are created:

```ts
MockProvider(ExampleService, {
  method1: () => of(example),
});
```

In this example, `method1` uses the explicit implementation while other unstubbed methods still throw. If you enable this only for a suite instead of globally, restore the previous configuration afterward with `ngMocks.autoSpy('reset')`.

This behavior applies to generated methods; ordinary property reads keep their normal mock behavior.
